### PR TITLE
abci++ spec: remove new_hashes and discuss transaction traceability

### PR DESCRIPTION
**Pull request by angbrav**
The PR includes a couple of changes to the spec to address #7663:
- Remove the field `new_hashes` from `TxRecord`. 
- Add text discussing the implications of removing `new_hashes`: the Application is responsible for transaction traceability.
- There is also a minor restructuring of the usage subsection of `PrepareProposal`. 

In my understanding (after talking to @sergio-mena), this is what was decided by the team. Let me know if that's not the case.

### CODE REVIEW

File: spec/abci++/abci++_methods_002_draft.md [hunk 1]
```diff
@@ -854,25 +842,3 @@ Most of the data structures used in ABCI are shared [common data structures](../
     |------------|-----------------------|------------------------------------------------------------------|--------------|
     | action     | [TxAction](#txaction) | What should Tendermint do with this transaction?                 | 1            |
     | tx         | bytes                 | Transaction contents                                             | 2            |
-    | new_hashes | repeated bytes        | List of hashes of successor transactions                         | 3            |
```
Comment (sergio-mena) [inline spec/abci++/abci++_methods_002_draft.md:857]: Please also remove this from the proto-bufs: proto/tendermint/abci/types.proto

File: spec/abci++/abci++_methods_002_draft.md [hunk 2]
```diff
@@ -312,17 +312,19 @@ From the App's perspective, they'll probably skip ProcessProposal
...
     * The header contains the height, timestamp, and more - it exactly matches the
       Tendermint block header.
-    * The Application can modify the transactions received in `RequestPrepareProposal` before sending
-      them in `ResponsePrepareProposal`. In that case, `ResponsePrepareProposal.modified_tx` is set to true.
-    * If `ResponsePrepareProposal.modified_tx` is false, then Tendermint will ignore the contents of
-      `ResponsePrepareProposal.tx_records`.
-    * If the Application modifies the transactions, the modified transactions MUST NOT exceed the configured maximum size,
-      contained in `RequestPrepareProposal.max_tx_bytes`.
+    * `RequestPrepareProposal` contains a preliminary set of transactions `txs` that Tendermint considers to be a good block proposal, called _raw block_. The Application can modify this set via `ResponsePrepareProposal.tx_records` (see [TxRecord](#txrecord)).
+        * In this case, the Applications should set `ResponsePrepareProposal.modified_tx` to true.
```
Comment (sergio-mena) [inline spec/abci++/abci++_methods_002_draft.md:320]:
```suggestion
        * In this case, the Application should set `ResponsePrepareProposal.modified_tx` to true.
```

File: spec/abci++/abci++_methods_002_draft.md [hunk 3]
```diff
@@ -312,17 +312,19 @@ From the App's perspective, they'll probably skip ProcessProposal
...
+        * If the Application modifies the set of transactions, the modified transactions MUST NOT exceed the configured maximum size `RequestPrepareProposal.max_tx_bytes`.
+    * If the Application does not modified the premilimnary set of transactions `txs`, then it sets `ResponsePrepareProposal.modified_tx` to false. In this case, Tendermint will ignore the contents of `ResponsePrepareProposal.tx_records`.
```
Comment (sergio-mena) [inline spec/abci++/abci++_methods_002_draft.md:327]:
```suggestion
    * If the Application does not modify the preliminary set of transactions `txs`, then it sets `ResponsePrepareProposal.modified_tx` to false. In this case, Tendermint will ignore the contents of `ResponsePrepareProposal.tx_records`.
```

File: spec/abci++/abci++_methods_002_draft.md [hunk 4]
```diff
@@ -403,7 +393,7 @@ and _p_'s _validValue_ is `nil`:
         * remove transactions (invalid) from the proposal and from the mempool - `TxAction = REMOVED`
         * remove transactions from the proposal but not from the mempool (effectively _delaying_ them) - the
           Application removes the transaction from the list
-        * modify transactions (e.g. aggregate them) - `TxAction = ADDED` followed by `TxAction = REMOVED`
+        * modify transactions (e.g. aggregate them) - `TxAction = ADDED` followed by `TxAction = REMOVED`. As explained above, this compromises client traceability.
```
Comment (sergio-mena) [inline spec/abci++/abci++_methods_002_draft.md:396]:
```suggestion
        * modify transactions (e.g. aggregate them) - `TxAction = ADDED` followed by `TxAction = REMOVED`. As explained above, this compromises client traceability, unless it is implemented at the Application level.
```

File: spec/abci++/abci++_methods_002_draft.md [hunk 5]
```diff
@@ -312,17 +312,19 @@ From the App's perspective, they'll probably skip ProcessProposal
...
+            * If the Application wants to add a new transaction, then the Application should include it in `tx_records` and _mark_ it as "ADD". In this case, Tendermint will add it to the mempool.
+        * The Application should be aware that removing and adding transactions may compromise _traceability_. Consider the following example: the Application transforms a client-submitted transaction `t1` into a second transaction `t2`, i.e., the Application asks Tendermint to remove `t1` and add `t2` to the mempool. If a client wants to eventually check what happened to `t1`, it will discover that `t_1` is not in the mempool or in a committed block, getting the wrong idea that `t_1` did not make it into a block. Note that `t_2` _will be_ in a committed block, but unless the Application tracks this information, no component will be aware of it. Thus, if the Application wants traceability, it is its responsability to support it. For instance, the Application could attach to a transformed transaction a list with the hashes of the transactions it derives from.
```
Comment (sergio-mena) [inline spec/abci++/abci++_methods_002_draft.md:325]: über-minor: I would add a line break before "Consider the following example:", making that sentence start a sub-bullet. But up to you!

**Review by sergio-mena — APPROVED** [review]
Thanks a lot @angbrav for this!

**Comment by sergio-mena** [comment]
Please turn this draft PR into a full PR

File: spec/abci++/abci++_methods_002_draft.md (hunk 5 above)
Comment (angbrav) [reply spec/abci++/abci++_methods_002_draft.md]: Done

**Pull request merged**